### PR TITLE
backend/feat: #755/get-driver-leaderboard

### PR DESCRIPTION
**Pull request by vikrantkm1997**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/LeaderBoard.hs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+newtype LeaderBoardRes = LeaderBoardRes
+  { driverList :: [DriversInfo]
+  }
+  deriving (Generic, Show, ToJSON, FromJSON, ToSchema)
+
+getDriverLeaderBoard :: (Esq.EsqDBFlow m r, Esq.EsqDBReplicaFlow m r, EncFlow m r, Redis.HedisFlow m r, CacheFlow m r) => Id Person -> Maybe Integer -> m LeaderBoardRes
+getDriverLeaderBoard personId mbLimit = do
+  person <- Esq.runInReplica $ QPerson.findById personId >>= fromMaybeM (PersonDoesNotExist personId.getId)
+  merchant <- QM.findById person.merchantId >>= fromMaybeM (MerchantNotFound person.merchantId.getId)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/LeaderBoard.hs:48]: Fetch config directly with fetching `merchant`. Anyways if merchant doesn't exist config will also be not there. So, we don't need to check for merchant separately.

File: Backend/dev/migrations/dynamic-offer-driver-app/0157-add-colums-to-driverStats.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+UPDATE atlas_driver_offer_bpp.driver_stats
+    SET total_rides = ride_counts.count,
+        total_distance = ride_counts.distance_sum
+    FROM (
+        SELECT driver_id, COUNT(*) AS count, SUM(chargeable_distance) AS distance_sum
+        FROM atlas_driver_offer_bpp.ride
+        WHERE status = 'COMPLETED'
+        GROUP BY driver_id
+    ) AS ride_counts
+    WHERE ride_counts.DriverId = rideStats.DriverId;
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0157-add-colums-to-driverStats.sql:15]: shouldn't it be `driver_id`?

File: Backend/dev/migrations/dynamic-offer-driver-app/0157-add-colums-to-driverStats.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+-- Add column in Transporter Configs
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0157-add-colums-to-driverStats.sql:17]: Add `not null` constraint also to `total_rides` and `total_distance` field after updated value.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/dev/migrations/dynamic-offer-driver-app/0157-add-colums-to-driverStats.sql (hunk 2 above)
Comment (0utkarsh) [reply Backend/dev/migrations/dynamic-offer-driver-app/0157-add-colums-to-driverStats.sql:15]: ride_stats too

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/LeaderBoard.hs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    Redis.get makeDriverLeaderBoardKey >>= \case
+      Nothing -> do
+        driversSortedList' <- Esq.runInReplica $ QDriverStats.getDriversSortedOrder mbLimit
+        drivers' <- forM (zip [1, 2 ..] driversSortedList') $ \(index, driver) -> do
+          person' <- Esq.runInReplica $ QPerson.findById (cast driver.driverId) >>= fromMaybeM (PersonDoesNotExist driver.driverId.getId)
+          fullName <- getPersonFullName person' >>= fromMaybeM (PersonFieldNotPresent "firstName")
+          pure DriversInfo {rank = index, name = fullName, totalRides = driver.totalRides, totalDistance = driver.totalDistance}
+        Redis.setExp makeDriverLeaderBoardKey drivers' $ getSeconds (fromMaybe 3600 config.driverLeaderBoardExpiry)
+        pure drivers'
+      Just a -> return a
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/LeaderBoard.hs:60]: mb rename `a`?

**Comment by ashishsingh101** [comment]
@vikrantkm1997 are we sending current driver info also in the response ?

**Review by hkmangla — APPROVED** [review]

**Pull request merged**